### PR TITLE
fix: Invalid URI when open invalid path

### PR DESCRIPTION
**Pull request by RussKie**
## Proposed changes

Attempt to open a repository at an invalid path (e.g. "C:" instead of "C:\\") would result in the following error:

    System.UriFormatException
      HResult=0x80131537
      Message=Invalid URI: The format of the URI could not be determined.
      Source=System
      StackTrace:
       at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
       at GitCommands.PathUtil.NormalizePath(String path)
       at GitCommands.GitModule..ctor(String workingDir)
       at GitUI.CommandsDialogs.BrowseDialog.FormOpenDirectory.LoadClick(Object sender, EventArgs e)

The fix adds checks around the user supplied path.

Relates to #6549
Relates to #6556

## Test methodology 

- manual
- add unit test

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs [hunk 1]
```diff
@@ -98,13 +98,14 @@ private void LoadClick(object sender, EventArgs e)
...
-            {
-                MessageBox.Show(this, _warningOpenFailed.Text, _warningOpenFailedCaption.Text);
+                if (_chosenModule.IsValidGitWorkingDir())
+                {
+                    ThreadHelper.JoinableTaskFactory.Run(() => RepositoryHistoryManager.Locals.AddAsMostRecentAsync(_chosenModule.WorkingDir));
+                    Close();
+                }
             }
+
+            MessageBox.Show(this, _warningOpenFailed.Text, _warningOpenFailedCaption.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
```
Comment (pmiossec) [inline GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs]: Won't this message be displayed every time now that it is no more in the else statement?

**Comment by RussKie** [comment]
Aye! Forgot the `return`.
Thank you

File: GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs]: fixed, thank you

**Comment by mstv** [comment]
It just eats the exception.
But the list of branches for URLs like "git@host:path.git" remains empty in the `Add submodule` dialog.
Then a random branch is chosen when you press `Add`.

File: GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs [hunk 2]
```diff
@@ -98,13 +98,15 @@ private void LoadClick(object sender, EventArgs e)
...
-                MessageBox.Show(this, _warningOpenFailed.Text, _warningOpenFailedCaption.Text);
+                if (_chosenModule.IsValidGitWorkingDir())
+                {
+                    ThreadHelper.JoinableTaskFactory.Run(() => RepositoryHistoryManager.Locals.AddAsMostRecentAsync(_chosenModule.WorkingDir));
+                    Close();
+                    return;
+                }
             }
+
+            MessageBox.Show(this, _warningOpenFailed.Text, _warningOpenFailedCaption.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs]: Suggestion: Not a valid Git directory

[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs]: You're absolutely right... 🤦‍♂ 
We can't add any tests (yet) until we add an abstraction for `MessageBox.Show`.

[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs]: Updated, please have a look

File: GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs [hunk 3]
```diff
@@ -95,16 +95,13 @@ public static GitModule OpenModule(IWin32Window owner, [CanBeNull] GitModule cur
         private void LoadClick(object sender, EventArgs e)
         {
             _NO_TRANSLATE_Directory.Text = _NO_TRANSLATE_Directory.Text.Trim();
-            if (Directory.Exists(_NO_TRANSLATE_Directory.Text))
+            if ((_chosenModule = OpenGitRepository(_NO_TRANSLATE_Directory.Text, RepositoryHistoryManager.Locals)) != null)
```
Comment (mstv) [inline GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs]: Just my 2 cents: I don't like assingments in `if` conditions.

Comment (RussKie) [reply GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs]: Nw, easy fix.

File: GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs [hunk 4]
```diff
@@ -149,5 +146,37 @@ private void _NO_TRANSLATE_Directory_TextChanged(object sender, EventArgs e)
                 folderGoUpButton.Enabled = false;
             }
         }
+
+        private GitModule OpenGitRepository([NotNull] string path, ILocalRepositoryManager localRepositoryManager)
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs]: done

**Pull request merged**